codice/ddf
# DDF-3121 Upgraded Karaf to 4.1.2

# DO NOT MERGE

#### What does this PR do?
This PR upgrades the DDF Karaf version from 4.1.1 to 4.1.2.
#### Who is reviewing it?
@vinamartin @brjeter @brendan-hofmann @paouelle
#### Select relevant component teams:
@codice/build @pklinef 
@codice/docs @ricklarsen 
@codice/security @stustison
#### Choose 2 committers to review/merge the PR. 
@clockard
@coyotesqrl
@kcwire
@figliold
@lessarderic
@pklinef
@stustison
#### How should this be tested?
- Full build
- Install
  - Install which each profile
  - Install with `?dev=true`, and restart
- Automatic Start on System Boot (see Documentation) on different OSs
- Config files
  - Install config files by dropping in the `/etc`
  - -`.config` files modified in this PR still work
- ` ./bin` scripts
  - Quit the karaf shell after `bin/start` and `bin/client` ([KARAF-5216](https://issues.apache.org/jira/browse/KARAF-5216) Exiting karaf shell, mess the bash shell)
  - ctrl-c while in bin/client ([KARAF-5218](https://issues.apache.org/jira/browse/KARAF-5218) bin/client exists when typing CTRL-C)
- Hardening Guide
- Commands/Console
  - `waitForReady`
  - `migration:export`, and exporting configurations from the Admin Console
  - `profile:`-
  - `app:`-
  - Use up and down arrows on the console command line ([KARAF-5176](https://issues.apache.org/jira/browse/KARAF-5176) Fix support for characters entered while executing a command)
  - `threads:tree` ([KARAF-5181](https://issues.apache.org/jira/browse/KARAF-5181) NPE while running "threads --tree" command from console)
  - timestamp not on by default in console ([KARAF-5276](https://issues.apache.org/jira/browse/KARAF-5276) Do not use right prompt by default)
- Logging
  - `log:tail`
  - `log:set <level> <logger>`
  - `log:get` ([KARAF-5091](https://issues.apache.org/jira/browse/KARAF-5091) log:get does not show correct level)
  - `log:list`, `log:list --help` ([KARAF-5182](https://issues.apache.org/jira/browse/KARAF-5182) Console command log:list returns "null")
  - Configure from `etc/org.ops4j.pax.logging.cfg`
  - XML-style log4j2 configuration (see documentation)
  - Location information is included in `log:tail` entries ([KARAF-5124](https://issues.apache.org/jira/browse/KARAF-5124) NPE when location information is included in console logging pattern)
  - no StoredWiringResolver NPEs in logs ([KARAF-5211](https://issues.apache.org/jira/browse/KARAF-5211) NPE in StoredWiringResolver if BundleEvent.UNRESOLVED handled before BundleEvent.RESOLVED event)
- [KARAF-5179](https://issues.apache.org/jira/browse/KARAF-5179) Setting the karaf.restart.jvm property to true causes system halt commands to behave as reboots
  - set `karaf.restart.jvm` to true
  - run the reboot command
  - it should shutdown and restart
- SSH ([KARAF-5223](https://issues.apache.org/jira/browse/KARAF-5223) "Error in initialization script" messages printed to the main console when clients connect through ssh)
  - intro running instance
  - into itests
#### Any background context you want to provide?
Generated tickets/known issues:
- [KARAF-5317](https://issues.apache.org/jira/browse/KARAF-5317) `Exception in thread "SIGWINCH handler" java.lang.UnsupportedOperationException` occurs when resizing the console while `log:tail` is run
- [KARAF-5311](https://issues.apache.org/jira/browse/KARAF-5311) NPE in karaf-maven-plugin when specifying descriptor by file url
- [FELIX-5306](https://issues.apache.org/jira/browse/FELIX-5306) Fileinstall can't read floats and doubles in `.config` files
- [KARAF-5332](https://issues.apache.org/jira/browse/KARAF-5332) `bin/stop` script fails when `KARAF_DEBUG` is set
- [KARAF-5320](https://issues.apache.org/jira/browse/KARAF-5320) Karaf Command Arguments escapes backslash characters @emanns95 
- [DDF-3258](https://codice.atlassian.net/browse/DDF-3258) Due to timing issues, DDF sometimes fails to install properly the first time
- [FELIX-5686](https://issues.apache.org/jira/browse/FELIX-5686) NPE in ConfigInstaller on ConfigDelete @brjeter 

This PR is blocked by:
- ~https://github.com/codice/ddf/pull/2291~
- ~https://github.com/codice/ddf/pull/2288~

This PR blocks:
- https://github.com/codice/alliance/pull/355

#### What are the relevant tickets?
[DDF-3121](https://codice.atlassian.net/browse/DDF-3121 )
#### Screenshots
[documentation.pdf](https://github.com/codice/ddf/files/1265431/documentation.pdf)
[jdocumentation.pdf](https://github.com/codice/ddf/files/1265432/jdocumentation.pdf)
![image](https://user-images.githubusercontent.com/8041246/29901086-e8dbbc62-8daa-11e7-886c-7cec6e14f8f3.png)
![image](https://user-images.githubusercontent.com/8041246/29901050-a761079c-8daa-11e7-825c-dac7621115ed.png)
![image](https://user-images.githubusercontent.com/8041246/29901059-c003fa7a-8daa-11e7-960d-0615b19de141.png)
#### Checklist:
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
